specs/progress.md

```diff
@@ -0,0 +1,150 @@
+# Progress
+
+## Overview
+
+*Progress* is used to display the length of time a process will take (referred to as a **determinate** state) or to represent an unspecified wait time (referred to as an **indeterminate** state). *Progress* is typically visually represented by a circular or linear animation. When the `value` attribute is passed the state is **determinate**, otherwise it is **indeterminate**.
+
+### Use Cases
+
+The *progress* component should be used for an element that displays the progress status for a task that takes a more than a few seconds or consists of several steps. Typical use cases include loading elements or status of a download. 
+  
+### Features
+
+A progress should allow the following attributes:
+- `value`, the value of the progress, if not passed the progress will be put into its "indeterminate" state.
+- `min`, the minimum value.
+- `max`, the maximum value.
+- `circular`, whether the progress renders circular or not.
+- `paused`, whether the progress is paused or not.
+
+### Prior Art/Examples
+- [FAST-DNA (React)](https://explore.fast.design/components/progress)
+- [Material UI](https://material-ui.com/components/progress/)
+- [Lightning Design / Linear](https://www.lightningdesignsystem.com/components/progress-bar/)
+- [Lightning Design / Circular](https://www.lightningdesignsystem.com/components/progress-ring/)
+- [Ant Design](https://ant.design/components/progress/)
+- [Atlassian / Linear](https://atlaskit.atlassian.com/packages/server/progress-bar)
+- [Atlassian / Circular](https://atlaskit.atlassian.com/packages/core/spinner)
+- [Windows (UWP)](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/progress-controls)
+
+---
+
+### API
+
+*Component name:*
+- `fast-progress`
+
+*Attributes:*
+- `value`: number
+- `min`: number
+- `max`: number
+- `circular`: boolean
+- `paused`: boolean
+
+*Slots:*
+- `indetermimate-linear`
+- `indetermimate-circular`
+
+### Anatomy and Appearance
+
+*Parts:*
+- determinate
+- indeterminate
+- circular-container
+- circular-background
+- circular-determinate
+- circular-indeterminate
```

Comment:
Is circular a completely different implementation? For something like a determinate progress bar, Ant has good examples of both linear and circular instances: https://ant.design/components/progress/

Reply:
I included that example above. I was wondering if it would simplify things if we just have 2 separate components `<fast-circular-progress/>` and `<fast-progress>`. Some libraries keep it as one others separate them out (Lightning Design - [Circular](https://www.lightningdesignsystem.com/components/progress-ring/) | [Linear](https://www.lightningdesignsystem.com/components/progress-bar/), Atlassian - [Circular](https://atlaskit.atlassian.com/packages/core/spinner) | [Linear](https://atlaskit.atlassian.com/packages/server/progress-bar))

Reply:
I'm actually inclined towards splitting these, the Lighting Design and Atlassian showcase a more extensible model which I think is preferable. At that point we could do Progress Bar and Progress Spinner or Progress Ring, both are being used by Fluent and UWP respectively.

Reply:
IMO we should be designing this component so that the visualization of progress can be extremely flexible at the template/styling level without changing the base code and in that sense any notion of "circular" should ideally be confined there.  Isn't circular just a style decision?

Reply:
See my above comment, circular is not just styling.

Reply:
With our current separation of component code, template, and styling, could this be one base `progress` class that deals with value and in/determinate mode and two templates, one for circular and one for linear? That seems most extensible as a consumer could create a dashboard style, a custom circular style, an illustration, etc.

I'm not sure if templates currently have the capability for this, but it seems ideal if the base component renders the container with the appropriate aria attributes, and the templates render the visual content.

This split seems like the most reuse of the basics without the need to duplicate many things like the parts or slots.

Reply:
I think the plan is to create a base component and then different templates. The folder structure will look something like this:

```
progress >
  circular-progress >
    circular-progress.stories.ts
    circular-progress.styles.ts
    circular-progress.template.ts
    index.ts
  radial-progress >
    linear-progress.stories.ts
    linear-progress.styles.ts
    linear-progress.template.ts
    index.ts
  base-progress.ts
  index.ts
```